### PR TITLE
Avoid adding generated items to the undo stack

### PR DESCRIPTION
**Pull request by miiizen**
Resolves: #17494 
Resolves: #18499
There's some refactoring done to adding/removing elements here.  `doUndoAdd/RemoveElement` prevents adding generated items to the undo stack.  This will hopefully reduce the number of crashes we get where an item is removed from the score by an undo command when it wasn't added by one.

### CODE REVIEW

**Review by cbjeukendrup — APPROVED** [review]

File: src/engraving/dom/undo.cpp [hunk 1]
```diff
@@ -795,6 +795,7 @@ void CloneVoice::redo(EditData*)
 
 AddElement::AddElement(EngravingItem* e)
 {
+    DO_ASSERT_X(!e->generated(), String("Generated item %1 passed to AddElement").arg(String(e->typeName())));
```
Comment (cbjeukendrup) [inline src/engraving/dom/undo.cpp]: Ah, this should be 
```suggestion
    DO_ASSERT_X(!e->generated(), String("Generated item %1 passed to AddElement").arg(String::fromUtf8(e->typeName())));
```
That will make the "without Qt" check happy.

(What's happening now, is that it creates a QString from the const char* and then creates a String from the QString.)

Comment (cbjeukendrup) [reply src/engraving/dom/undo.cpp]: Oh, you just did... 
Somehow I keep commenting on GitHub pages that I apparently had to refresh 😅

File: src/engraving/rendering/dev/measurelayout.cpp [hunk 2]
```diff
@@ -1375,7 +1375,7 @@ void MeasureLayout::layoutMeasureNumber(Measure* m, LayoutContext& ctx)
                 if (t->generated()) {
                     ctx.mutDom().removeElement(t);
                 } else {
-                    ctx.mutDom().undo(new RemoveElement(t));
+                    ctx.mutDom().doUndoRemoveElement(t);
                 }
```
Comment (cbjeukendrup) [inline src/engraving/rendering/dev/measurelayout.cpp:1379]: It looks like some redundancy has now arisen here. Same a few lines below.

**Comment by cbjeukendrup** [comment]
I just came across https://github.com/musescore/MuseScore/pull/20331#discussion_r1420149649; doesn't this PR suffer from the same to some extent? (because of the `RemoveElement` constructor)

File: src/engraving/dom/undo.cpp [hunk 3]
```diff
@@ -795,6 +795,7 @@ void CloneVoice::redo(EditData*)
 
 AddElement::AddElement(EngravingItem* e)
 {
+    DO_ASSERT_X(!e->generated(), String("Generated item %1 passed to AddElement").arg(String::fromAscii(e->typeName())));
```
Comment (cbjeukendrup) [inline src/engraving/dom/undo.cpp:798]: In this case I think 
```suggestion
    DO_ASSERT_X(!e->generated(), String(u"Generated item %1 passed to AddElement").arg(String::fromAscii(e->typeName())));
```

**Comment by miiizen** [comment]
Yeah I definaitely agree with what Roman says.  I suppose as long as the item types mentioned in `RemoveElement` are never generated it's ok and AFAIK they aren't, but feels like quite a big assumption to make.  I'd stress that this PR is one to fix the crash quickly for 4.2.

Also, we already have four ways of adding/removing an element (the methods in Score, undo and adding directly to measure/segment) - my feeling is that any refactoring we do should cut this number down, not add a fifth!  This PR probably makes the design problem worse unfortunately.

File: src/engraving/dom/edit.cpp [hunk 4]
```diff
@@ -6427,17 +6437,7 @@ void Score::undoRemoveElement(EngravingItem* element, bool removeLinked)
             if (e->explicitParent() && e->explicitParent()->isSystem()) {
                 e->setParent(0);   // systems will be regenerated upon redo, so detach
             }
-        }
-    }
-
-    if (element->isMeasureRepeat()) {
```
Comment (RomanPudashkin) [inline src/engraving/dom/edit.cpp:6433]: Why did you move that piece of code?

Comment (miiizen) [reply src/engraving/dom/edit.cpp:6433]: I deleted the elements immediately in `doUndoRemoveElement` before I found `deleteLater`.  This is unnecessary now, I'll move it back.

**Review by RomanPudashkin — APPROVED** [review]
